File: javascript/dsa/containerWithMostWater.js
var maxArea = function (height) {
    let l = 0;
    let r = height.length - 1;
    let maxArea = 0;
    while (l < r) {
        let currArea = Math.min(height[l], height[r]) * (r - l);
        maxArea = Math.max(currArea, maxArea);

        if (height[l] < height[r]) {
            l += 1;
        } else {
            r -= 1;
        }
    }
    return maxArea;
};

console.log(maxArea([1, 8, 6, 2, 5, 4, 8, 3, 7]));

File: two-pointers/containerWithMostWater.py
class Solution:
    def maxArea(self, height) -> int:
        # Time complexity : O(n)
        # Space complexity: O(1)
        l, r = 0, len(height) - 1
        res = 0
        while l < r:
            area = (r - l) * min(height[l], height[r])
            res = max(res, area)
            if height[l] < height[r]:
                l += 1
            else:
                r -= 1
        return res


a = Solution()
b = a.maxArea([1, 8, 6, 2, 5, 4, 8, 3, 7])
print(b)
